gprofiler/perf.py

```diff
@@ -42,42 +46,41 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.stop()
 
-    def _run_perf(self, filename_base: str, dwarf=False):
-        parsed_path = os.path.join(self._storage_dir, f"{filename_base}.parsed")
-
+    def _run_perf(self, dwarf: bool = False) -> str:
         buildid_args = ["--buildid-dir", PERF_BUILDID_DIR]
 
         with NamedTemporaryFile(dir=self._storage_dir) as record_file:
             args = ["-F", str(self._frequency), "-a", "-g", "-o", record_file.name]
             if dwarf:
-                args += ["--call-graph", "dwarf"]
+                args += ["--call-graph", f"dwarf,{self._dwarf_stack_size}"]
             run_process(
                 [resource_path("perf")] + buildid_args + ["record"] + args + ["--", "sleep", str(self._duration)],
                 stop_event=self._stop_event,
             )
-            with open(parsed_path, "w") as f:
-                run_process(
-                    [resource_path("perf")] + buildid_args + ["script", "-F", "+pid", "-i", record_file.name], stdout=f
-                )
-            return parsed_path
+            perf_script_result = run_process(
+                [resource_path("perf")] + buildid_args + ["script", "-F", "+pid", "-i", record_file.name],
+                suppress_log=True,
+            )
+            return perf_script_result.stdout.decode('utf8')
 
-    def snapshot(self) -> Iterable[Mapping[str, str]]:
+    def snapshot(self) -> Tuple[ProcessToStackSampleCounters, ProcessIdToNameMapping]:
         free_disk = psutil.disk_usage(self._storage_dir).free
         if free_disk < 4 * 1024 * 1024:
             raise Exception(f"Free disk space: {free_disk}kb. Skipping perf!")
 
         logger.info("Running global perf...")
-        record_path = self._run_perf("global")
-        logger.info("Finished running global perf")
-        return parse_perf_script(Path(record_path).read_text())
-
-        # TODO: run dwarf in parallel, after supporting it in merge.py
-        # Alternatively: fix golang dwarf problems and the run just dwarf
+        if not self._fp_perf:
+            return merge_global_perfs(None, self._run_perf(dwarf=True))
+        if not self._dwarf_perf:
+            return merge_global_perfs(self._run_perf(dwarf=False), None)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+            # We are running 2 perfs in parallel - one with DWARF and one with FP, and then we merge their results.
+            # This improves the results from software that is compiled without frame pointers, like some Go software.
```

Comment:
It's actually the other way around - Go has FP but doesn't work with DWARF for whatever reason which we believe is a bug in DWARF unwinding in `perf` or one of the libraries it's based on.